dftimewolf/lib/processors/gcp_cloud_resource_tree.py

```diff
@@ -0,0 +1,1032 @@
+              disk_resource.parent.name, disk_resource.parent.type,
+              disk_resource.parent.zone, disk_resource.parent.project_id)
+          if matched_resource:
+            disk_resource.parent = matched_resource
+
+        if disk_resource and not resource.parent:
+          resource.parent = disk_resource
+
+        if disk_resource.id:
+          self.resources_dict[disk_resource.id] = disk_resource
+
+    if resource.parent and not resource.parent.id:
+      matched_resource = self._FindResource(resource.parent.name,
+                                            resource.parent.type,
+                                            resource.parent.zone,
+                                            resource.parent.project_id)
+      if matched_resource:
+        resource.parent = matched_resource
+
+  def _BuildResourcesParentRelationships(self) -> None:
+    """Builds parent relationship for all resources."""
+    # Using resource_keys because self.resources_dict changes during the loop
+    resource_keys = list(self.resources_dict.keys())
+    for resource_key in resource_keys:
+      resource = self.resources_dict.get(resource_key)
+      if resource:
+        resource.parent = self._GetResourceParentTree(resource)
+
+  def _SearchForDeletedResource(
+      self, resource: gcp_crt_helper.Resource,
+      start_timestamp: datetime) -> Optional[gcp_crt_helper.Resource]:
+    """Searches for deleted resource in GCP Logs.
+
+    Args:
+      resource: resource to search for.
+      start_timestamp: the initial point of time to start the search
+
+    Return:
+      Found resource or None
+    """
+    if not resource or not start_timestamp:
+      return None
+
+    if resource.project_id != self.project_id:
+      return None
+
+    while start_timestamp > (datetime.now(timezone.utc) - timedelta(days=400)):
+
+      end_timestamp = start_timestamp + timedelta(minutes=20)
+      start_timestamp = start_timestamp - timedelta(days=30)
+
+      log_messages = self._GetLogMessages(resource.project_id, start_timestamp,
+                                          end_timestamp, resource.id)
+      self._ParseLogMessages(log_messages)
+
+      matched_resource = self._FindResource(resource.name, resource.type)
+      if matched_resource:
+        if matched_resource.deletion_timestamp and matched_resource.creation_timestamp:  # pylint: disable=line-too-long
+          return matched_resource
+
+    return None
+
+  def _RetrieveListOfDisks(
+      self, project_id: str) -> Dict[str, gcp_crt_helper.Resource]:
+    """Retrieves list of disks in a project.
+
+    Args:
+      project_id: Project Id to retrieve the list of disks for
+
+    Returns:
+      Dict of disks
+    """
+    result: Dict[str, gcp_crt_helper.Resource] = {}
+
+    # Using beta version of the API because v1 did not have important
+    # information when creating this script
+    compute_api_client = gcp_common.CreateService('compute', 'beta')
```

Comment:
you create a new client object and do full auth handshake with every call to CreateService, you can cache the compute client once in the setup method (if no creds this will fails which is another reason to do this in setup and avoid failing late), for example either:
- make it instance variable, cache client in setup 
`self.compute_api_client = gcp_common.CreateService('compute', 'v1')`
- or use (it defaults to v1):
sleft.compute_api_client = gcp_common.GoogleCloudComputeClient(project_id).GceApi()

Reply:
I will create one instance in `_GetListOfResources` and pass it to each of the resource. As for the check, I have a preflight check in the recipe to check if the user has access to the project.